test/python/dialects/indexing/dialect.py

```diff
@@ -1306,15 +1306,88 @@ def testDynSliceScatter():
+      b = a + a
+      return b
+
+  module.operation.verify()
+  pm = PassManager.parse('builtin.module(cse,canonicalize)')
+  pm.run(module.operation)
+  print(module)
+
+
 # CHECK-LABEL: TEST: testForLoopSugar
 @run
 def testForLoopSugar():
   f32 = F32Type.get()
   with mlir_mod_ctx() as module:
-    ten = Tensor.empty((7, 22, 330, 4400), f32)
 
-    for i in scf_range(0, 10):
-      y = 2 * i
+    @func.FuncOp.from_py_func(*[])
+    def test_matmul():
+      A = Tensor.empty((10, 20), f32)
+      B = Tensor.empty((20, 30), f32)
+      C = Tensor.empty((10, 30), f32)
+      # tile size
+      ts = 5
+      for i, (C,), r1 in scf_range(0, 10, ts, iter_args=[C]):
+        for j, (C,), r2 in scf_range(0, 30, ts, iter_args=[C]):
+          for k, (C,), r3 in scf_range(0, 20, ts, iter_args=[C]):
+            C = Tensor(C)
+
+            a = A[i:i + ts, k:k + ts, is_slice]
+            b = B[k:k + ts, j:j + ts, is_slice]
+            C[i:i + ts, j:j + ts, is_slice] += a @ b
+
+            scf_yield(C)
+          scf_yield(r3)
+        scf_yield(r2)
+
+      return r1
 
   module.operation.verify()
+  pm = PassManager.parse(
+      'builtin.module(cse,func.func(gather-to-extract-slice,scatter-to-insert-slice))'
+  )
+  pm.run(module.operation)
+  # CHECK: module {
+  # CHECK:   func.func @test_matmul() -> tensor<10x30xf32> {
+  # CHECK:     %cst = arith.constant 0.000000e+00 : f32
+  # CHECK:     %c1 = arith.constant 1 : index
+  # CHECK:     %c20 = arith.constant 20 : index
+  # CHECK:     %c30 = arith.constant 30 : index
+  # CHECK:     %c5 = arith.constant 5 : index
+  # CHECK:     %c10 = arith.constant 10 : index
+  # CHECK:     %c0 = arith.constant 0 : index
+  # CHECK:     %0 = tensor.empty() : tensor<10x20xf32>
+  # CHECK:     %1 = tensor.empty() : tensor<20x30xf32>
+  # CHECK:     %2 = tensor.empty() : tensor<10x30xf32>
+  # CHECK:     %3 = scf.for %arg0 = %c0 to %c10 step %c5 iter_args(%arg1 = %2) -> (tensor<10x30xf32>) {
+  # CHECK:       %4 = scf.for %arg2 = %c0 to %c30 step %c5 iter_args(%arg3 = %arg1) -> (tensor<10x30xf32>) {
+  # CHECK:         %5 = scf.for %arg4 = %c0 to %c20 step %c5 iter_args(%arg5 = %arg3) -> (tensor<10x30xf32>) {
+  # CHECK:           %extracted_slice = tensor.extract_slice %0[%arg0, %arg4] [%c5, %c5] [%c1, %c1] : tensor<10x20xf32> to tensor<?x?xf32>
+  # CHECK:           %extracted_slice_0 = tensor.extract_slice %1[%arg4, %arg2] [%c5, %c5] [%c1, %c1] : tensor<20x30xf32> to tensor<?x?xf32>
+  # CHECK:           %extracted_slice_1 = tensor.extract_slice %arg5[%arg0, %arg2] [%c5, %c5] [%c1, %c1] : tensor<10x30xf32> to tensor<?x?xf32>
+  # CHECK:           %6 = tensor.empty(%c5, %c5) : tensor<?x?xf32>
+  # CHECK:           %7 = linalg.fill ins(%cst : f32) outs(%6 : tensor<?x?xf32>) -> tensor<?x?xf32>
+  # CHECK:           %8 = linalg.matmul {cast = #linalg.type_fn<cast_signed>} ins(%extracted_slice, %extracted_slice_0 : tensor<?x?xf32>, tensor<?x?xf32>) outs(%7 : tensor<?x?xf32>) -> tensor<?x?xf32>
```

Comment:
This should take extracted_slice_1 as the outs parameter and then you don't need the extra empty, fill and add.

Reply:
add such a test case

Reply:
i take it back - maybe it's not so questionable after all :p (implemented `c += a @ b` with no fill.